src/interface_rpc_url_requirer.py

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/python3
+
+"""RPC url interface (providers side)."""
+
+from service_args import ServiceArgs
+from ops.framework import Object
+from ops.charm import RelationChangedEvent, RelationDepartedEvent
+import utils
+
+class RpcUrlRequirer(Object):
+    """RPC url requirer interface."""
+
+    def __init__(self, charm, relation_name):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self.framework.observe(
+            charm.on[relation_name].relation_changed, self._on_relation_changed
+        )
+        self.framework.observe(
+            charm.on[relation_name].relation_departed, self._on_relation_departed
+        )
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """This event is used to receive the http rpc url from another client."""
+
+        if not event.unit in event.relation.data:
+            event.defer()
+            return
+
+        # The --relay-chain-rpc-urls option currently only supports ws, hence using ws_url and not rpc_url.
+        try:
+            ws_url = event.relation.data[event.unit]["ws_url"]
+        except KeyError:
+            event.defer()
+            return
+        # Storing the unitname+relation_id is a workaround because the relation data is already removed before the departed hook is called.
+        # This is to know which url to remove.
+        # This "bug" (unclear if it's a bug or a removed feature) is reported to the Juju team.
+        dict_key = event.unit.name + ':' + str(event.relation.id)
+        self._charm._stored.relay_rpc_urls[dict_key] = ws_url
```

Comment:
So this is the work around where we need to store info about the unit since the data is gone in the departed hook.